hammers/query.py

```diff
@@ -284,6 +284,55 @@ def update_orphan_resource_providers(db):
     '''
     return db.query(sql, no_rows=True)
 
+@query
+def get_advance_reservations(db):
+    """Get all advance reservations created by any user"""
+    sql = '''\
+    SELECT lu.name AS user_name, u.extra AS user_extra, l.name AS lease_name, l.id AS lease_id, p.name AS project_name, l.start_date AS start_date 
+    FROM blazar.leases AS l
+    JOIN keystone.user AS u ON l.user_id = u.id 
+    JOIN keystone.local_user AS lu ON u.id = lu.user_id 
+    JOIN keystone.project AS p ON l.project_id = p.id 
+    WHERE l.start_date > UTC_TIMESTAMP()
+    '''
+    
+    return db.query(sql, limit=None)
+
+@query
+def get_idle_leases(db, hours):
+    """
+    Get all leases that have been idle for more than a given number hours.
+    An idle lease is defined as an ongoing lease without active instance and without lease update operation.
+    """
+    sql = '''\
+    SELECT lu.name AS user_name, u.extra AS user_extra, s.lease_name AS lease_name, s.lease_id AS lease_id, p.name AS project_name
+    FROM (
+        SELECT l.id AS lease_id, l.name AS lease_name, l.user_id, l.project_id
+        FROM blazar.leases AS l
+        JOIN blazar.reservations AS r ON l.id = r.lease_id
+        JOIN blazar.computehost_allocations AS ca ON ca.reservation_id = r.id
+        JOIN blazar.computehosts ch on ca.compute_host_id=ch.id
+        LEFT JOIN nova.instances AS i on i.node = ch.hypervisor_hostname
+        WHERE l.deleted_at IS NULL
+          AND i.created_at >= l.start_date
+          AND l.start_date < UTC_TIMESTAMP()
+          AND l.end_date > UTC_TIMESTAMP()
+          AND TIMESTAMPDIFF(HOUR, l.updated_at, UTC_TIMESTAMP()) >= %s
+          AND TIMESTAMPDIFF(HOUR, r.updated_at, UTC_TIMESTAMP()) >= %s
+        GROUP BY l.id
+        HAVING COUNT(DISTINCT i.uuid) = 0
+          OR (
+             COUNT(DISTINCT i.uuid) > 0
+             AND SUM(ISNULL(i.deleted_at)) = 0
+             AND TIMESTAMPDIFF(HOUR, MAX(i.updated_at), UTC_TIMESTAMP()) >= %s
+          )
```

Comment:
Can you please explain what the part between OR ( ) is meant to select?

Reply:
no active instance under the lease and the latest touch of any instance is greater than a threshold.